src/transaction/log_reader.hpp

```diff
@@ -86,54 +86,56 @@ class log_reader final
 
     /* equivalent to LOG_READ_ALIGN
      */
-    void align ();
+    inline void align ();
 
     /* equivalent to LOG_READ_ADD_ALIGN
      */
-    void add_align (size_t size);
+    inline void add_align (size_t size);
 
     template <typename T>
-    void add_align ();
+    inline void add_align ();
 
     /* equivalent to LOG_READ_ADVANCE_WHEN_DOESNT_FIT
      */
-    void advance_when_does_not_fit (size_t size);
+    inline void advance_when_does_not_fit (size_t size);
 
     /* returns whether the supplied lengths is contained in the currently
      * loaded log page (also considering the current offset)
      */
-    bool does_fit_in_current_page (size_t size) const;
+    inline bool does_fit_in_current_page (size_t size) const;
 
     /* copy from log into externally supplied buffer
      * also advancing the - internally kept - page pointer if needed
      */
-    void copy_from_log (char *dest, size_t length);
+    inline void copy_from_log (char *dest, size_t length);
 
     /*
      * TODO: somehow this function, add_align and advance_when_does_not_fit
      * have the same core functionality and could be combined
      */
-    int skip (size_t size);
+    inline int skip (size_t size);
 
   private:
-    const char *get_cptr () const;
+    inline const char *get_cptr () const;
 
-    int fetch_page_force_use (THREAD_ENTRY *const thread_p);
+    inline int fetch_page_force_use (THREAD_ENTRY *const thread_p);
 
   private:
     log_lsa m_lsa = NULL_LSA;
     log_page *m_page;
     char m_area_buffer[IO_MAX_PAGE_SIZE + DOUBLE_ALIGNMENT];
 };
 
-void LOG_READ_ALIGN (THREAD_ENTRY *thread_p, LOG_LSA *lsa, LOG_PAGE *log_pgptr);
-void LOG_READ_ADD_ALIGN (THREAD_ENTRY *thread_p, size_t add, LOG_LSA *lsa, LOG_PAGE *log_pgptr);
-void LOG_READ_ADVANCE_WHEN_DOESNT_FIT (THREAD_ENTRY *thread_p, size_t length, LOG_LSA *lsa,
-				       LOG_PAGE *log_pgptr);
+inline void LOG_READ_ALIGN (THREAD_ENTRY *thread_p, LOG_LSA *lsa, LOG_PAGE *log_pgptr);
+inline void LOG_READ_ADD_ALIGN (THREAD_ENTRY *thread_p, size_t add, LOG_LSA *lsa, LOG_PAGE *log_pgptr);
+inline void LOG_READ_ADVANCE_WHEN_DOESNT_FIT (THREAD_ENTRY *thread_p, size_t length, LOG_LSA *lsa,
+    LOG_PAGE *log_pgptr);
 
 
 /* implementation
  */
+#include "log_impl.h"
+#include "thread_manager.hpp"
 
 template <typename T>
 const typename std::remove_reference<T>::type *log_reader::reinterpret_cptr () const
```

Comment:
all log_reader implementations should be in the same order as the declarations

Reply:
keeping a mirrored implements and declaration is a great effort with little benefit. devs are usually relying on IDE navigation to go from declarations to the corresponding implementations.

but the template functions are moved at the end, because it is usual to start the implementation with the constructor/destructor functions.